apps/aecore/src/aec_sync.erl

```diff
@@ -1117,3 +1136,89 @@ pp_sync_task(ST = #sync_task{}) ->
                 , {#sync_task.pending, fun(X) -> lists:map(PPPoolF, X) end}
                 , {#sync_task.workers, fun pp_worker/1}
                 ]).
+
+process_infos(Infos) ->
+    {Responded, Failed0} =
+        lists:foldl(
+            fun({error, Err}, {RAccum, FAccum}) ->
+                {RAccum, [Err | FAccum]};
+               ({ok, #{} = NodeInfo }, {RAccum, FAccum}) ->
+                {[NodeInfo | RAccum], FAccum}
+            end,
+            {[], []},
+            Infos),
+      Failed =
+          lists:foldl(
+              fun(Err, Accum) ->
+                  maps:put(Err, maps:get(Err, Accum, 0) + 1, Accum)
+              end,
+              #{},
+              Failed0),
+      NoneIfEmpty =
+          fun(M) ->
+              case map_size(M) =:= 0 of
+                  true -> none;
+                  false -> M 
+              end
+          end,
+      Aggr =
+          fun(Key) ->
+              Res =
+                  lists:foldl(
+                      fun(#{Key := NodeVsn}, Accum) ->
+                          maps:put(NodeVsn, maps:get(NodeVsn, Accum, 0) + 1, Accum)
+                      end,
+                      #{},
+                      Responded),
+              NoneIfEmpty(Res)
+          end,
+      #{ versions => Aggr(node_version)
+        , os       => Aggr(os)
+        , failed   => NoneIfEmpty(Failed)}.
+
+collect_infos(Timeout) ->
+    ConnectedPeers = aec_peers:connected_peers(),
+    TimerOffset = 500, %% time overhead for this process' timeout
+    Fun =
+        fun(PeerInfo) ->
+            PeerId = aec_peer:id(PeerInfo),
+            %% this timeouts in Timeout milliseconds but there is a second
+            %% outher timeout in the gen_server:call to the connection itself
+            %% so we call the aec_peer_connection:get_node_info/2 with 
+            %% Timeout - 2000 - TimerOffset
+            aec_peer_connection:get_node_info(PeerId, Timeout - 2000 -
+                                              TimerOffset)
+        end,
+    pmap(Fun, ConnectedPeers, Timeout + 2000 + TimerOffset).
+
+pmap(Fun, L, Timeout) ->
+    Parent = self(),
+    lists:foreach(
+        fun(E) ->
+            spawn(fun() ->
+                      Child = self(),
+                      spawn(fun() ->
+                                Res = Fun(E),
+                                Child ! {pmap_result_inner, Res}
+                            end),
+                      Result =
+                          receive
+                              {pmap_result_inner, R} -> R
+                          after Timeout -> {error, request_timeout}
+                          end,
+                      Parent ! {pmap_result, Result}
+                  end)
+        end,
+        L),
+    pmap_gather(length(L), []).
```

Comment:
One way to write such a pmap (apologies for syntax errors):

```erlang
Pids = [spawn_monitor(fun() -> exit({ok, PerformTask(E)} end) || E <- L],
pmap_gather(Pids, []).

pmap_gather([], Acc) ->
    Acc;
pmap_gather([{Pid, MRef}|Pids], Acc) ->
    receive
        {'DOWN', MRef, process, Pid, Res} ->
            case Res of
                {ok, GoodRes} -> pmap_gather(Pids, [GoodRes | Acc]);
                Error -> do_something_clever()
            end
    end.
```

This also captures crashes, and makes it possible to tell the difference.